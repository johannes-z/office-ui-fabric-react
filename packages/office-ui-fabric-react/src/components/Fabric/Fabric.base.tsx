import * as React from 'react';
import {
  Customizer,
  getNativeProps,
  on,
  divProperties,
  classNamesFunction,
  getWindow,
  getDocument,
  isDirectionalKeyCode,
  memoizeFunction
} from '../../Utilities';
import { getStyles } from './Fabric.styles';
import { IFabricProps, IFabricStyleProps, IFabricStyles } from './Fabric.types';
import { IProcessedStyleSet } from '@uifabric/merge-styles';
import { ITheme } from '../../Styling';

const getClassNames = classNamesFunction<IFabricStyleProps, IFabricStyles>();
const getRTLTheme = memoizeFunction((theme: ITheme, isRTL: boolean) => ({ ...theme, rtl: isRTL }));

export class FabricBase extends React.Component<
  IFabricProps,
  {
    isFocusVisible: boolean;
  }
> {
  private _rootElement = React.createRef<HTMLDivElement>();
  private _disposables: (() => void)[] = [];
  private _removeClassNameFromBody?: () => void = undefined;

  constructor(props: IFabricProps) {
    super(props);
    this.state = { isFocusVisible: false };
  }

  public render() {
    const { as: Root = 'div', theme, dir } = this.props;
    const classNames = this._getClassNames();
    const divProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(this.props, divProperties);
    const isRTL = dir === 'rtl';
    let renderedContent = <Root {...divProps} className={classNames.root} ref={this._rootElement} />;

    // Expose an rtl based theme if dir is specified and it doesn't agree with the theme setting.
    if (dir !== undefined && theme && (theme.rtl === undefined || theme.rtl !== isRTL)) {
      renderedContent = <Customizer settings={{ theme: getRTLTheme(theme, isRTL) }}>{renderedContent}</Customizer>;
    }

    return renderedContent;
  }

  public componentDidMount(): void {
    const win = getWindow(this._rootElement.current);
    if (win) {
      this._disposables.push(on(win, 'mousedown', this._onMouseDown, true), on(win, 'keydown', this._onKeyDown, true));
    }
    this._addClassNameToBody();
  }

  public componentWillUnmount(): void {
    this._disposables.forEach((dispose: () => void) => dispose());
    if (this._removeClassNameFromBody) {
      this._removeClassNameFromBody();
    }
  }

  private _getClassNames(): IProcessedStyleSet<IFabricStyles> {
    const { className, theme, applyTheme } = this.props;
    const classNames = getClassNames(getStyles, {
      theme: theme!,
      applyTheme: applyTheme,
      className,
      isFocusVisible: this.state.isFocusVisible
    });
    return classNames;
  }

  private _addClassNameToBody(): void {
    if (this.props.applyThemeToBody) {
      const classNames = this._getClassNames();
      const currentDoc = getDocument(this._rootElement.current);
      if (currentDoc) {
        currentDoc.body.classList.add(classNames.bodyThemed);
        this._removeClassNameFromBody = () => {
          currentDoc.body.classList.remove(classNames.bodyThemed);
        };
      }
    }
  }

  private _onMouseDown = (ev: MouseEvent): void => {
    this.setState({ isFocusVisible: false });
  };

  private _onKeyDown = (ev: KeyboardEvent): void => {
    if (isDirectionalKeyCode(ev.which)) {
      this.setState({ isFocusVisible: true });
    }
  };
}
