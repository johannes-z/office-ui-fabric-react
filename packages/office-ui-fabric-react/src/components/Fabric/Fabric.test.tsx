import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Fabric } from './Fabric';
import { Checkbox } from '../../Checkbox';

import { mount } from 'enzyme';

describe('Fabric', () => {
  it('renders a Fabric component correctly', () => {
    const component = renderer.create(<Fabric>test</Fabric>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a Fabric component with applyTheme correctly', () => {
    const component = renderer.create(<Fabric applyTheme>test</Fabric>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a Fabric component in RTL', () => {
    const component = renderer.create(
      <div>
        <Checkbox label="I am a default Checkbox" />
        <Fabric dir="ltr">
          <Checkbox label="I am in ltr" />

          <Fabric dir="rtl">
            <Checkbox label="I am in rtl, inside of ltr" />

            <Fabric dir="ltr">
              <Checkbox label="I am in ltr, inside of rtl, inside of ltr" />
            </Fabric>
          </Fabric>
        </Fabric>
      </div>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders as a span using the "as" prop', () => {
    const component = renderer.create(<Fabric as="span" />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a Fabric component with applyThemeToBody correctly', () => {
    const wrapper = mount(<Fabric applyThemeToBody>test</Fabric>);
    expect(wrapper.getDOMNode().getElementsByClassName('bodyThemed')).toBeTruthy();
  });
});
