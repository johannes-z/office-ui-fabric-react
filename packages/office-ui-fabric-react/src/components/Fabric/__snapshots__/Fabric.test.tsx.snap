// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fabric renders a Fabric component correctly 1`] = `
<div
  className=
      ms-Fabric
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        color: #323130;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
      }
      & button {
        font-family: inherit;
      }
      & input {
        font-family: inherit;
      }
      & textarea {
        font-family: inherit;
      }
>
  test
</div>
`;

exports[`Fabric renders a Fabric component in RTL 1`] = `
<div>
  <div
    className=
        ms-Checkbox
        is-enabled
        {
          display: flex;
          position: relative;
        }
        &:hover .ms-Checkbox-checkbox {
          border-color: #323130;
        }
        @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkbox {
          border-color: Highlight;
        }
        &:focus .ms-Checkbox-checkbox {
          border-color: #323130;
        }
        &:hover .ms-Checkbox-checkmark {
          color: #605e5c;
          opacity: 1;
        }
        @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkmark {
          color: Highlight;
        }
        &:hover .ms-Checkbox-text {
          color: #201f1e;
        }
        &:focus .ms-Checkbox-text {
          color: #201f1e;
        }
  >
    <input
      aria-checked="false"
      aria-label="I am a default Checkbox"
      checked={false}
      className=

          {
            background: none;
            opacity: 0;
            position: absolute;
          }
          .ms-Fabric--isFocusVisible &:focus + label::before {
            outline-offset: 2px;
            outline: 1px solid #605e5c;
          }
          @media screen and (-ms-high-contrast: active){.ms-Fabric--isFocusVisible &:focus + label::before {
            outline: 1px solid ActiveBorder;
          }
      id="checkbox-0"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    <label
      className=
          ms-Checkbox-label
          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            align-items: flex-start;
            cursor: pointer;
            display: flex;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 14px;
            font-weight: 400;
            position: relative;
            text-align: left;
            user-select: none;
          }
          &::before {
            bottom: 0px;
            content: "";
            left: 0px;
            pointer-events: none;
            position: absolute;
            right: 0px;
            top: 0px;
          }
      htmlFor="checkbox-0"
    >
      <div
        className=
            ms-Checkbox-checkbox
            {
              align-items: center;
              border-radius: 2px;
              border: 1px solid #323130;
              box-sizing: border-box;
              display: flex;
              flex-shrink: 0;
              height: 20px;
              justify-content: center;
              margin-right: 4px;
              overflow: hidden;
              position: relative;
              transition-duration: 200ms;
              transition-property: background, border, border-color;
              transition-timing-function: cubic-bezier(.4, 0, .23, 1);
              width: 20px;
            }
      >
        <i
          aria-hidden={true}
          className=
              ms-Checkbox-checkmark
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                color: #ffffff;
                display: inline-block;
                font-family: "FabricMDL2Icons";
                font-style: normal;
                font-weight: normal;
                opacity: 0;
                speak: none;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: Window;
              }
          data-icon-name="CheckMark"
        >
          
        </i>
      </div>
      <span
        aria-hidden="true"
        className=
            ms-Checkbox-text
            {
              color: #323130;
              font-size: 14px;
              line-height: 20px;
              margin-left: 4px;
            }
      >
        I am a default Checkbox
      </span>
    </label>
  </div>
  <div
    className=
        ms-Fabric
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #323130;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
        }
        & button {
          font-family: inherit;
        }
        & input {
          font-family: inherit;
        }
        & textarea {
          font-family: inherit;
        }
    dir="ltr"
  >
    <div
      className=
          ms-Checkbox
          is-enabled
          {
            display: flex;
            position: relative;
          }
          &:hover .ms-Checkbox-checkbox {
            border-color: #323130;
          }
          @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkbox {
            border-color: Highlight;
          }
          &:focus .ms-Checkbox-checkbox {
            border-color: #323130;
          }
          &:hover .ms-Checkbox-checkmark {
            color: #605e5c;
            opacity: 1;
          }
          @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkmark {
            color: Highlight;
          }
          &:hover .ms-Checkbox-text {
            color: #201f1e;
          }
          &:focus .ms-Checkbox-text {
            color: #201f1e;
          }
    >
      <input
        aria-checked="false"
        aria-label="I am in ltr"
        checked={false}
        className=

            {
              background: none;
              opacity: 0;
              position: absolute;
            }
            .ms-Fabric--isFocusVisible &:focus + label::before {
              outline-offset: 2px;
              outline: 1px solid #605e5c;
            }
            @media screen and (-ms-high-contrast: active){.ms-Fabric--isFocusVisible &:focus + label::before {
              outline: 1px solid ActiveBorder;
            }
        id="checkbox-1"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
      />
      <label
        className=
            ms-Checkbox-label
            {
              -moz-osx-font-smoothing: grayscale;
              -webkit-font-smoothing: antialiased;
              align-items: flex-start;
              cursor: pointer;
              display: flex;
              font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
              font-size: 14px;
              font-weight: 400;
              position: relative;
              text-align: left;
              user-select: none;
            }
            &::before {
              bottom: 0px;
              content: "";
              left: 0px;
              pointer-events: none;
              position: absolute;
              right: 0px;
              top: 0px;
            }
        htmlFor="checkbox-1"
      >
        <div
          className=
              ms-Checkbox-checkbox
              {
                align-items: center;
                border-radius: 2px;
                border: 1px solid #323130;
                box-sizing: border-box;
                display: flex;
                flex-shrink: 0;
                height: 20px;
                justify-content: center;
                margin-right: 4px;
                overflow: hidden;
                position: relative;
                transition-duration: 200ms;
                transition-property: background, border, border-color;
                transition-timing-function: cubic-bezier(.4, 0, .23, 1);
                width: 20px;
              }
        >
          <i
            aria-hidden={true}
            className=
                ms-Checkbox-checkmark
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  color: #ffffff;
                  display: inline-block;
                  font-family: "FabricMDL2Icons";
                  font-style: normal;
                  font-weight: normal;
                  opacity: 0;
                  speak: none;
                }
                @media screen and (-ms-high-contrast: active){& {
                  -ms-high-contrast-adjust: none;
                  color: Window;
                }
            data-icon-name="CheckMark"
          >
            
          </i>
        </div>
        <span
          aria-hidden="true"
          className=
              ms-Checkbox-text
              {
                color: #323130;
                font-size: 14px;
                line-height: 20px;
                margin-left: 4px;
              }
        >
          I am in ltr
        </span>
      </label>
    </div>
    <div
      className=
          ms-Fabric
          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: #323130;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 14px;
            font-weight: 400;
          }
          & button {
            font-family: inherit;
          }
          & input {
            font-family: inherit;
          }
          & textarea {
            font-family: inherit;
          }
      dir="rtl"
    >
      <div
        className=
            ms-Checkbox
            is-enabled
            {
              display: flex;
              position: relative;
            }
            &:hover .ms-Checkbox-checkbox {
              border-color: #323130;
            }
            @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkbox {
              border-color: Highlight;
            }
            &:focus .ms-Checkbox-checkbox {
              border-color: #323130;
            }
            &:hover .ms-Checkbox-checkmark {
              color: #605e5c;
              opacity: 1;
            }
            @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkmark {
              color: Highlight;
            }
            &:hover .ms-Checkbox-text {
              color: #201f1e;
            }
            &:focus .ms-Checkbox-text {
              color: #201f1e;
            }
      >
        <input
          aria-checked="false"
          aria-label="I am in rtl, inside of ltr"
          checked={false}
          className=

              {
                background: none;
                opacity: 0;
                position: absolute;
              }
              .ms-Fabric--isFocusVisible &:focus + label::before {
                outline-offset: 2px;
                outline: 1px solid #605e5c;
              }
              @media screen and (-ms-high-contrast: active){.ms-Fabric--isFocusVisible &:focus + label::before {
                outline: 1px solid ActiveBorder;
              }
          id="checkbox-2"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="checkbox"
        />
        <label
          className=
              ms-Checkbox-label
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                align-items: flex-start;
                cursor: pointer;
                display: flex;
                font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 14px;
                font-weight: 400;
                position: relative;
                text-align: right;
                user-select: none;
              }
              &::before {
                bottom: 0px;
                content: "";
                left: 0px;
                pointer-events: none;
                position: absolute;
                right: 0px;
                top: 0px;
              }
          htmlFor="checkbox-2"
        >
          <div
            className=
                ms-Checkbox-checkbox
                {
                  align-items: center;
                  border-radius: 2px;
                  border: 1px solid #323130;
                  box-sizing: border-box;
                  display: flex;
                  flex-shrink: 0;
                  height: 20px;
                  justify-content: center;
                  margin-left: 4px;
                  overflow: hidden;
                  position: relative;
                  transition-duration: 200ms;
                  transition-property: background, border, border-color;
                  transition-timing-function: cubic-bezier(.4, 0, .23, 1);
                  width: 20px;
                }
          >
            <i
              aria-hidden={true}
              className=
                  ms-Checkbox-checkmark
                  {
                    -moz-osx-font-smoothing: grayscale;
                    -webkit-font-smoothing: antialiased;
                    color: #ffffff;
                    display: inline-block;
                    font-family: "FabricMDL2Icons";
                    font-style: normal;
                    font-weight: normal;
                    opacity: 0;
                    speak: none;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    -ms-high-contrast-adjust: none;
                    color: Window;
                  }
              data-icon-name="CheckMark"
            >
              
            </i>
          </div>
          <span
            aria-hidden="true"
            className=
                ms-Checkbox-text
                {
                  color: #323130;
                  font-size: 14px;
                  line-height: 20px;
                  margin-right: 4px;
                }
          >
            I am in rtl, inside of ltr
          </span>
        </label>
      </div>
      <div
        className=
            ms-Fabric
            {
              -moz-osx-font-smoothing: grayscale;
              -webkit-font-smoothing: antialiased;
              color: #323130;
              font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
              font-size: 14px;
              font-weight: 400;
            }
            & button {
              font-family: inherit;
            }
            & input {
              font-family: inherit;
            }
            & textarea {
              font-family: inherit;
            }
        dir="ltr"
      >
        <div
          className=
              ms-Checkbox
              is-enabled
              {
                display: flex;
                position: relative;
              }
              &:hover .ms-Checkbox-checkbox {
                border-color: #323130;
              }
              @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkbox {
                border-color: Highlight;
              }
              &:focus .ms-Checkbox-checkbox {
                border-color: #323130;
              }
              &:hover .ms-Checkbox-checkmark {
                color: #605e5c;
                opacity: 1;
              }
              @media screen and (-ms-high-contrast: active){&:hover .ms-Checkbox-checkmark {
                color: Highlight;
              }
              &:hover .ms-Checkbox-text {
                color: #201f1e;
              }
              &:focus .ms-Checkbox-text {
                color: #201f1e;
              }
        >
          <input
            aria-checked="false"
            aria-label="I am in ltr, inside of rtl, inside of ltr"
            checked={false}
            className=

                {
                  background: none;
                  opacity: 0;
                  position: absolute;
                }
                .ms-Fabric--isFocusVisible &:focus + label::before {
                  outline-offset: 2px;
                  outline: 1px solid #605e5c;
                }
                @media screen and (-ms-high-contrast: active){.ms-Fabric--isFocusVisible &:focus + label::before {
                  outline: 1px solid ActiveBorder;
                }
            id="checkbox-3"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
          />
          <label
            className=
                ms-Checkbox-label
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  align-items: flex-start;
                  cursor: pointer;
                  display: flex;
                  font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                  position: relative;
                  text-align: left;
                  user-select: none;
                }
                &::before {
                  bottom: 0px;
                  content: "";
                  left: 0px;
                  pointer-events: none;
                  position: absolute;
                  right: 0px;
                  top: 0px;
                }
            htmlFor="checkbox-3"
          >
            <div
              className=
                  ms-Checkbox-checkbox
                  {
                    align-items: center;
                    border-radius: 2px;
                    border: 1px solid #323130;
                    box-sizing: border-box;
                    display: flex;
                    flex-shrink: 0;
                    height: 20px;
                    justify-content: center;
                    margin-right: 4px;
                    overflow: hidden;
                    position: relative;
                    transition-duration: 200ms;
                    transition-property: background, border, border-color;
                    transition-timing-function: cubic-bezier(.4, 0, .23, 1);
                    width: 20px;
                  }
            >
              <i
                aria-hidden={true}
                className=
                    ms-Checkbox-checkmark
                    {
                      -moz-osx-font-smoothing: grayscale;
                      -webkit-font-smoothing: antialiased;
                      color: #ffffff;
                      display: inline-block;
                      font-family: "FabricMDL2Icons";
                      font-style: normal;
                      font-weight: normal;
                      opacity: 0;
                      speak: none;
                    }
                    @media screen and (-ms-high-contrast: active){& {
                      -ms-high-contrast-adjust: none;
                      color: Window;
                    }
                data-icon-name="CheckMark"
              >
                
              </i>
            </div>
            <span
              aria-hidden="true"
              className=
                  ms-Checkbox-text
                  {
                    color: #323130;
                    font-size: 14px;
                    line-height: 20px;
                    margin-left: 4px;
                  }
            >
              I am in ltr, inside of rtl, inside of ltr
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Fabric renders a Fabric component with applyTheme correctly 1`] = `
<div
  className=
      ms-Fabric
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background-color: #ffffff;
        color: #323130;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
      }
      & button {
        font-family: inherit;
      }
      & input {
        font-family: inherit;
      }
      & textarea {
        font-family: inherit;
      }
>
  test
</div>
`;

exports[`Fabric renders as a span using the "as" prop 1`] = `
<span
  className=
      ms-Fabric
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        color: #323130;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
      }
      & button {
        font-family: inherit;
      }
      & input {
        font-family: inherit;
      }
      & textarea {
        font-family: inherit;
      }
/>
`;
